File: test_automation.js
/**
 * Automated Test Suite for Master Data - Products & Services
 * Test ID: TC-MD-PS-001
 * Framework: Playwright
 * Base URL: https://app-twenty5ipe-lm-dev.cfapps.us10.hana.ondemand.com/
 */

const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

// Configuration
const BASE_URL = 'https://app-twenty5ipe-lm-dev.cfapps.us10.hana.ondemand.com/';
const HEADLESS = false; // Set to true for CI/CD environments
const SCREENSHOT_DIR = './test-screenshots';
const REPORT_FILE = './test-report.json';

// Test report structure
const testReport = {
  timestamp: new Date().toISOString(),
  baseUrl: BASE_URL,
  results: [],
  summary: {
    totalTests: 0,
    passed: 0,
    failed: 0,
    skipped: 0
  }
};

// Utility function to create screenshot directory
function setupScreenshotDirectory() {
  if (!fs.existsSync(SCREENSHOT_DIR)) {
    fs.mkdirSync(SCREENSHOT_DIR, { recursive: true });
  }
}

// Utility function to take screenshots
async function takeScreenshot(page, testName, stepName) {
  const fileName = `${testName}-${stepName.replace(/\s+/g, '_')}-${Date.now()}.png`;
  const filePath = path.join(SCREENSHOT_DIR, fileName);
  await page.screenshot({ path: filePath });
  return filePath;
}

// Utility function to log test results
function logTestResult(testId, testName, stepName, passed, message, screenshot = null) {
  const result = {
    testId,
    testName,
    stepName,
    passed,
    message,
    timestamp: new Date().toISOString(),
    screenshot
  };
  testReport.results.push(result);
  testReport.summary.totalTests++;
  if (passed) {
    testReport.summary.passed++;
  } else {
    testReport.summary.failed++;
  }
  console.log(`[${passed ? '✓ PASS' : '✗ FAIL'}] ${testId} - ${stepName}: ${message}`);
}

// Test Case 1: Navigate to Master Data Tab and Verify Sub-tabs
async function testCase1(page) {
  const testId = 'TC-MD-PS-001.1';
  const testName = 'Navigate to Master Data Tab';

  try {
    console.log(`\n▶ Running ${testId}: ${testName}`);

    // Step 1: Navigate to the application
    console.log('  Step 1: Navigating to application...');
    await page.goto(BASE_URL, { waitUntil: 'networkidle' });
    let screenshot = await takeScreenshot(page, testId, 'initial-load');
    logTestResult(testId, testName, 'Page loaded', true, 'Application loaded successfully', screenshot);

    // Step 2: Look for Master Data tab/menu
    console.log('  Step 2: Looking for Master Data tab...');
    try {
      // Try multiple selectors for Master Data
      const masterDataSelectors = [
        'text=Master Data',
        '[aria-label*="Master Data"]',
        'button:has-text("Master Data")',
        'a:has-text("Master Data")',
        '.nav-item:has-text("Master Data")',
        '[class*="masterdata"]',
        '[class*="master-data"]'
      ];

      let masterDataElement = null;
      for (const selector of masterDataSelectors) {
        try {
          masterDataElement = await page.$(selector);
          if (masterDataElement) break;
        } catch (e) {
          // Continue to next selector
        }
      }

      if (masterDataElement) {
        console.log('  Clicking Master Data element...');
        await masterDataElement.click();
        await page.waitForTimeout(1000);
        screenshot = await takeScreenshot(page, testId, 'master-data-clicked');
        logTestResult(testId, testName, 'Master Data Tab Clicked', true, 'Master Data tab clicked successfully', screenshot);
      } else {
        logTestResult(testId, testName, 'Master Data Tab Found', false, 'Could not locate Master Data tab', screenshot);
        return false;
      }

      // Step 3: Verify sub-tabs are displayed
      console.log('  Step 3: Verifying sub-tabs...');
      const subtabSelectors = [
        'text=Products & Services',
        '[class*="subtab"]',
        '.sub-menu',
        '[role="menuitem"]'
      ];

      let subtabsVisible = false;
      for (const selector of subtabSelectors) {
        try {
          const element = await page.$(selector);
          if (element) {
            subtabsVisible = true;
            break;
          }
        } catch (e) {
          // Continue
        }
      }

      screenshot = await takeScreenshot(page, testId, 'subtabs-verification');
      if (subtabsVisible) {
        logTestResult(testId, testName, 'Sub-tabs Visible', true, 'Sub-tabs are displayed', screenshot);
      } else {
        logTestResult(testId, testName, 'Sub-tabs Visible', false, 'Sub-tabs not found', screenshot);
      }

      return subtabsVisible;
    } catch (error) {
      screenshot = await takeScreenshot(page, testId, 'error');
      logTestResult(testId, testName, 'Test Execution', false, `Error: ${error.message}`, screenshot);
      return false;
    }
  } catch (error) {
    console.error(`Test ${testId} failed:`, error);
    return false;
  }
}

// Test Case 2: Navigate to Products & Services Page
async function testCase2(page) {
  const testId = 'TC-MD-PS-001.2';
  const testName = 'Navigate to Products & Services Page';

  try {
    console.log(`\n▶ Running ${testId}: ${testName}`);

    // Step 1: Click on Products & Services
    console.log('  Step 1: Clicking Products & Services...');
    const psSelectors = [
      'text=Products & Services',
      'text=Products / Services',
      '[aria-label*="Products"]',
      'a:has-text("Products & Services")',
      'button:has-text("Products & Services")'
    ];

    let clicked = false;
    for (const selector of psSelectors) {
      try {
        const element = await page.$(selector);
        if (element) {
          await element.click();
          clicked = true;
          break;
        }
      } catch (e) {
        // Continue
      }
    }

    if (!clicked) {
      let screenshot = await takeScreenshot(page, testId, 'products-not-found');
      logTestResult(testId, testName, 'Products & Services Click', false, 'Could not locate Products & Services option', screenshot);
      return false;
    }

    // Wait for navigation
    await page.waitForTimeout(2000);
    let screenshot = await takeScreenshot(page, testId, 'products-page');
    logTestResult(testId, testName, 'Page Navigated', true, 'Navigated to Products & Services page', screenshot);

    // Step 2: Verify URL
    console.log('  Step 2: Verifying URL...');
    const currentUrl = page.url();
    const urlValid = currentUrl.includes('product') || currentUrl.includes('service');
    logTestResult(testId, testName, 'URL Verification', urlValid, `Current URL: ${currentUrl}`, screenshot);

    return true;
  } catch (error) {
    console.error(`Test ${testId} failed:`, error);
    let screenshot = await takeScreenshot(page, testId, 'error');
    logTestResult(testId, testName, 'Test Execution', false, `Error: ${error.message}`, screenshot);
    return false;
  }
}

// Test Case 3: Verify New Button Options
async function testCase3(page) {
  const testId = 'TC-MD-PS-001.3';
  const testName = 'Verify New Button Options';

  try {
    console.log(`\n▶ Running ${testId}: ${testName}`);

    // Step 1: Find New button
    console.log('  Step 1: Locating New button...');
    const newButtonSelectors = [
      'button:has-text("New")',
      'text=New',
      '[aria-label*="New"]',
      '[class*="new-btn"]',
      '.btn-new'
    ];

    let newButton = null;
    for (const selector of newButtonSelectors) {
      try {
        newButton = await page.$(selector);
        if (newButton) break;
      } catch (e) {
        // Continue
      }
    }

    if (!newButton) {
      let screenshot = await takeScreenshot(page, testId, 'new-button-not-found');
      logTestResult(testId, testName, 'New Button Found', false, 'Could not locate New button', screenshot);
      return false;
    }

    // Step 2: Hover over New button
    console.log('  Step 2: Hovering over New button...');
    await newButton.hover();
    await page.waitForTimeout(500);

    let screenshot = await takeScreenshot(page, testId, 'new-button-hovered');
    logTestResult(testId, testName, 'New Button Hovered', true, 'Hovered over New button', screenshot);

    // Step 3: Verify dropdown options
    console.log('  Step 3: Verifying dropdown options...');
    const expectedOptions = [
      'Create from Template',
      'Create from Type',
      'Create Template',
      'Edit Template'
    ];

    let foundOptions = [];
    for (const option of expectedOptions) {
      try {
        const element = await page.$(`text=${option}`);
        if (element) {
          foundOptions.push(option);
        }
      } catch (e) {
        // Continue
      }
    }

    screenshot = await takeScreenshot(page, testId, 'dropdown-options');
    const allOptionsFound = foundOptions.length === expectedOptions.length;
    logTestResult(
      testId,
      testName,
      'Dropdown Options Verified',
      allOptionsFound,
      `Found ${foundOptions.length}/${expectedOptions.length} options: ${foundOptions.join(', ')}`,
      screenshot
    );

    return allOptionsFound;
  } catch (error) {
    console.error(`Test ${testId} failed:`, error);
    let screenshot = await takeScreenshot(page, testId, 'error');
    logTestResult(testId, testName, 'Test Execution', false, `Error: ${error.message}`, screenshot);
    return false;
  }
}

// Test Case 4: Verify Create from Type Additional Options
async function testCase4(page) {
  const testId = 'TC-MD-PS-001.4';
  const testName = 'Verify Additional Options on Create from Type';

  try {
    console.log(`\n▶ Running ${testId}: ${testName}`);

    // Step 1: Hover over Create from Type
    console.log('  Step 1: Hovering over Create from Type...');
    const createFromTypeSelectors = [
      'text=Create from Type',
      '[class*="create-from-type"]'
    ];

    let createFromTypeOption = null;
    for (const selector of createFromTypeSelectors) {
      try {
        createFromTypeOption = await page.$(selector);
        if (createFromTypeOption) break;
      } catch (e) {
        // Continue
      }
    }

    if (!createFromTypeOption) {
      let screenshot = await takeScreenshot(page, testId, 'create-from-type-not-found');
      logTestResult(testId, testName, 'Create from Type Found', false, 'Could not locate Create from Type option', screenshot);
      return false;
    }

    await createFromTypeOption.hover();
    await page.waitForTimeout(500);

    let screenshot = await takeScreenshot(page, testId, 'create-from-type-hovered');
    logTestResult(testId, testName, 'Create from Type Hovered', true, 'Hovered over Create from Type option', screenshot);

    // Step 2: Verify Purchased Part option
    console.log('  Step 2: Verifying Purchased Part option...');
    const purchasedPartSelectors = [
      'text=Purchased Part',
      '[class*="purchased-part"]'
    ];

    let purchasedPartFound = false;
    for (const selector of purchasedPartSelectors) {
      try {
        const element = await page.$(selector);
        if (element) {
          purchasedPartFound = true;
          break;
        }
      } catch (e) {
        // Continue
      }
    }

    screenshot = await takeScreenshot(page, testId, 'submenu-options');
    logTestResult(
      testId,
      testName,
      'Purchased Part Option Found',
      purchasedPartFound,
      purchasedPartFound ? 'Purchased Part option is visible' : 'Purchased Part option not found',
      screenshot
    );

    return purchasedPartFound;
  } catch (error) {
    console.error(`Test ${testId} failed:`, error);
    let screenshot = await takeScreenshot(page, testId, 'error');
    logTestResult(testId, testName, 'Test Execution', false, `Error: ${error.message}`, screenshot);
    return false;
  }
}

// Test Case 5: Create Purchased Part and Verify Sub-tabs
async function testCase5(page) {
  const testId = 'TC-MD-PS-001.5';
  const testName = 'Create Purchased Part and Verify Sub-tabs';

  try {
    console.log(`\n▶ Running ${testId}: ${testName}`);

    // Step 1: Click Purchased Part
    console.log('  Step 1: Clicking Purchased Part...');
    const purchasedPartSelectors = [
      'text=Purchased Part',
      '[class*="purchased-part"]'
    ];

    let clicked = false;
    for (const selector of purchasedPartSelectors) {
      try {
        const element = await page.$(selector);
        if (element) {
          await element.click();
          clicked = true;
          break;
        }
      } catch (e) {
        // Continue
      }
    }

    if (!clicked) {
      let screenshot = await takeScreenshot(page, testId, 'purchased-part-not-found');
      logTestResult(testId, testName, 'Purchased Part Clicked', false, 'Could not locate or click Purchased Part', screenshot);
      return false;
    }

    // Wait for page navigation
    await page.waitForTimeout(2000);
    let screenshot = await takeScreenshot(page, testId, 'create-page-loaded');
    logTestResult(testId, testName, 'Create Page Loaded', true, 'Navigated to Create Purchased Part page', screenshot);

    // Step 2: Verify required sub-tabs
    console.log('  Step 2: Verifying sub-tabs...');
    const requiredSubtabs = [
      'Key Info',
      'Product Management',
      'Planning & Estimating',
      'Cost & Prices',
      'Supplier Sources',
      'Comments & Files',
      'Texts'
    ];

    let foundSubtabs = [];
    for (const subtab of requiredSubtabs) {
      try {
        const element = await page.$(`text=${subtab}`);
        if (element) {
          foundSubtabs.push(subtab);
        }
      } catch (e) {
        // Continue
      }
    }

    screenshot = await takeScreenshot(page, testId, 'subtabs-verification');
    const allSubtabsFound = foundSubtabs.length === requiredSubtabs.length;
    logTestResult(
      testId,
      testName,
      'Sub-tabs Verified',
      allSubtabsFound,
      `Found ${foundSubtabs.length}/${requiredSubtabs.length} sub-tabs: ${foundSubtabs.join(', ')}`,
      screenshot
    );

    // Step 3: Verify each sub-tab is accessible
    if (allSubtabsFound) {
      console.log('  Step 3: Verifying sub-tabs are clickable...');
      let allClickable = true;
      for (const subtab of requiredSubtabs) {
        try {
          const element = await page.$(`text=${subtab}`);
          if (element) {
            // Just verify visibility, don't necessarily click to avoid side effects
            const isVisible = await element.isVisible();
            if (!isVisible) {
              allClickable = false;
            }
          }
        } catch (e) {
          allClickable = false;
        }
      }

      screenshot = await takeScreenshot(page, testId, 'subtabs-clickable');
      logTestResult(
        testId,
        testName,
        'Sub-tabs Accessible',
        allClickable,
        allClickable ? 'All sub-tabs are visible and accessible' : 'Some sub-tabs are not accessible',
        screenshot
      );
    }

    return allSubtabsFound;
  } catch (error) {
    console.error(`Test ${testId} failed:`, error);
    let screenshot = await takeScreenshot(page, testId, 'error');
    logTestResult(testId, testName, 'Test Execution', false, `Error: ${error.message}`, screenshot);
    return false;
  }
}

// Main test execution
async function runTests() {
  console.log('========================================');
  console.log('Master Data - Products & Services Tests');
  console.log('========================================');
  console.log(`Base URL: ${BASE_URL}`);
  console.log(`Timestamp: ${new Date().toISOString()}\n`);

  setupScreenshotDirectory();

  let browser;
  try {
    // Launch browser
    browser = await chromium.launch({ headless: HEADLESS });
    const context = await browser.newContext();
    const page = await context.newPage();

    // Set viewport
    await page.setViewportSize({ width: 1920, height: 1080 });

    // Run test cases
    const test1Result = await testCase1(page);
    const test2Result = await testCase2(page);
    const test3Result = await testCase3(page);
    const test4Result = await testCase4(page);
    const test5Result = await testCase5(page);

    // Generate report
    console.log('\n========================================');
    console.log('Test Summary');
    console.log('========================================');
    console.log(`Total Tests: ${testReport.summary.totalTests}`);
    console.log(`Passed: ${testReport.summary.passed}`);
    console.log(`Failed: ${testReport.summary.failed}`);
    console.log(`Pass Rate: ${((testReport.summary.passed / testReport.summary.totalTests) * 100).toFixed(2)}%`);

    // Save report
    fs.writeFileSync(REPORT_FILE, JSON.stringify(testReport, null, 2));
    console.log(`\nTest report saved to: ${REPORT_FILE}`);
    console.log(`Screenshots saved to: ${SCREENSHOT_DIR}`);

    await context.close();
    await browser.close();
  } catch (error) {
    console.error('Test execution failed:', error);
    if (browser) {
      await browser.close();
    }
    process.exit(1);
  }
}

// Run tests
runTests().catch(error => {
  console.error('Fatal error:', error);
  process.exit(1);
});

File: test_automation.py
#!/usr/bin/env python3
"""
Automated Test Suite for Master Data - Products & Services
Test ID: TC-MD-PS-001
Framework: Selenium WebDriver
Base URL: https://app-twenty5ipe-lm-dev.cfapps.us10.hana.ondemand.com/
"""

import time
import json
import os
from datetime import datetime
from pathlib import Path

from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.action_chains import ActionChains
from selenium.common.exceptions import TimeoutException, NoSuchElementException

# Configuration
BASE_URL = 'https://app-twenty5ipe-lm-dev.cfapps.us10.hana.ondemand.com/'
SCREENSHOT_DIR = './test-screenshots'
REPORT_FILE = './test-report.json'
WAIT_TIMEOUT = 10

# Test report structure
test_report = {
    'timestamp': datetime.now().isoformat(),
    'baseUrl': BASE_URL,
    'results': [],
    'summary': {
        'totalTests': 0,
        'passed': 0,
        'failed': 0,
        'skipped': 0
    }
}


class TestAutomation:
    def __init__(self):
        """Initialize the test automation class"""
        self.setup_directories()
        self.driver = None
        self.wait = None
        self.actions = None

    def setup_directories(self):
        """Create necessary directories for screenshots and reports"""
        Path(SCREENSHOT_DIR).mkdir(exist_ok=True)

    def init_driver(self):
        """Initialize WebDriver"""
        print("Initializing WebDriver...")
        options = webdriver.ChromeOptions()
        # options.add_argument('--headless')  # Comment out to see browser
        options.add_argument('--no-sandbox')
        options.add_argument('--disable-dev-shm-usage')
        options.add_argument('--disable-blink-features=AutomationControlled')
        options.add_argument('user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36')

        self.driver = webdriver.Chrome(options=options)
        self.wait = WebDriverWait(self.driver, WAIT_TIMEOUT)
        self.actions = ActionChains(self.driver)
        self.driver.set_window_size(1920, 1080)

    def take_screenshot(self, test_name, step_name):
        """Take a screenshot and save it"""
        timestamp = int(time.time() * 1000)
        filename = f"{test_name}-{step_name.replace(' ', '_')}-{timestamp}.png"
        filepath = os.path.join(SCREENSHOT_DIR, filename)
        self.driver.save_screenshot(filepath)
        return filepath

    def log_result(self, test_id, test_name, step_name, passed, message, screenshot=None):
        """Log test result"""
        result = {
            'testId': test_id,
            'testName': test_name,
            'stepName': step_name,
            'passed': passed,
            'message': message,
            'timestamp': datetime.now().isoformat(),
            'screenshot': screenshot
        }
        test_report['results'].append(result)
        test_report['summary']['totalTests'] += 1
        if passed:
            test_report['summary']['passed'] += 1
            print(f"[✓ PASS] {test_id} - {step_name}: {message}")
        else:
            test_report['summary']['failed'] += 1
            print(f"[✗ FAIL] {test_id} - {step_name}: {message}")

    def find_element_by_text(self, text, tag_name='*'):
        """Find element by visible text"""
        try:
            xpath = f"//{tag_name}[contains(text(), '{text}')]"
            return self.driver.find_element(By.XPATH, xpath)
        except NoSuchElementException:
            return None

    def find_elements_by_text(self, text_list):
        """Find multiple elements by text"""
        found = []
        for text in text_list:
            element = self.find_element_by_text(text)
            if element:
                found.append(text)
        return found

    def hover_element(self, element):
        """Hover over an element"""
        self.actions.move_to_element(element).perform()
        time.sleep(0.5)

    def test_case_1(self):
        """Test Case 1: Navigate to Master Data Tab and Verify Sub-tabs"""
        test_id = 'TC-MD-PS-001.1'
        test_name = 'Navigate to Master Data Tab'

        print(f"\n▶ Running {test_id}: {test_name}")

        try:
            # Step 1: Navigate to the application
            print("  Step 1: Navigating to application...")
            self.driver.get(BASE_URL)
            time.sleep(2)
            screenshot = self.take_screenshot(test_id, 'initial-load')
            self.log_result(test_id, test_name, 'Page loaded', True, 'Application loaded successfully', screenshot)

            # Step 2: Look for Master Data tab/menu
            print("  Step 2: Looking for Master Data tab...")
            master_data_element = self.find_element_by_text('Master Data')

            if master_data_element:
                master_data_element.click()
                time.sleep(1)
                screenshot = self.take_screenshot(test_id, 'master-data-clicked')
                self.log_result(test_id, test_name, 'Master Data Tab Clicked', True,
                              'Master Data tab clicked successfully', screenshot)
            else:
                screenshot = self.take_screenshot(test_id, 'master-data-not-found')
                self.log_result(test_id, test_name, 'Master Data Tab Found', False,
                              'Could not locate Master Data tab', screenshot)
                return False

            # Step 3: Verify sub-tabs are displayed
            print("  Step 3: Verifying sub-tabs...")
            products_element = self.find_element_by_text('Products & Services')
            screenshot = self.take_screenshot(test_id, 'subtabs-verification')

            if products_element:
                self.log_result(test_id, test_name, 'Sub-tabs Visible', True,
                              'Sub-tabs are displayed', screenshot)
                return True
            else:
                self.log_result(test_id, test_name, 'Sub-tabs Visible', False,
                              'Sub-tabs not found', screenshot)
                return False

        except Exception as error:
            screenshot = self.take_screenshot(test_id, 'error')
            self.log_result(test_id, test_name, 'Test Execution', False,
                          f'Error: {str(error)}', screenshot)
            return False

    def test_case_2(self):
        """Test Case 2: Navigate to Products & Services Page"""
        test_id = 'TC-MD-PS-001.2'
        test_name = 'Navigate to Products & Services Page'

        print(f"\n▶ Running {test_id}: {test_name}")

        try:
            # Step 1: Click on Products & Services
            print("  Step 1: Clicking Products & Services...")
            ps_element = self.find_element_by_text('Products & Services')

            if not ps_element:
                ps_element = self.find_element_by_text('Products / Services')

            if ps_element:
                ps_element.click()
                time.sleep(2)
                screenshot = self.take_screenshot(test_id, 'products-page')
                self.log_result(test_id, test_name, 'Page Navigated', True,
                              'Navigated to Products & Services page', screenshot)
            else:
                screenshot = self.take_screenshot(test_id, 'products-not-found')
                self.log_result(test_id, test_name, 'Products & Services Click', False,
                              'Could not locate Products & Services option', screenshot)
                return False

            # Step 2: Verify URL
            print("  Step 2: Verifying URL...")
            current_url = self.driver.current_url
            url_valid = 'product' in current_url.lower() or 'service' in current_url.lower()
            self.log_result(test_id, test_name, 'URL Verification', url_valid,
                          f'Current URL: {current_url}', screenshot)

            return True

        except Exception as error:
            screenshot = self.take_screenshot(test_id, 'error')
            self.log_result(test_id, test_name, 'Test Execution', False,
                          f'Error: {str(error)}', screenshot)
            return False

    def test_case_3(self):
        """Test Case 3: Verify New Button Options"""
        test_id = 'TC-MD-PS-001.3'
        test_name = 'Verify New Button Options'

        print(f"\n▶ Running {test_id}: {test_name}")

        try:
            # Step 1: Find New button
            print("  Step 1: Locating New button...")
            new_button = self.find_element_by_text('New', 'button')

            if not new_button:
                new_button = self.find_element_by_text('New')

            if not new_button:
                screenshot = self.take_screenshot(test_id, 'new-button-not-found')
                self.log_result(test_id, test_name, 'New Button Found', False,
                              'Could not locate New button', screenshot)
                return False

            # Step 2: Hover over New button
            print("  Step 2: Hovering over New button...")
            self.hover_element(new_button)
            screenshot = self.take_screenshot(test_id, 'new-button-hovered')
            self.log_result(test_id, test_name, 'New Button Hovered', True,
                          'Hovered over New button', screenshot)

            # Step 3: Verify dropdown options
            print("  Step 3: Verifying dropdown options...")
            expected_options = [
                'Create from Template',
                'Create from Type',
                'Create Template',
                'Edit Template'
            ]

            found_options = self.find_elements_by_text(expected_options)
            screenshot = self.take_screenshot(test_id, 'dropdown-options')

            all_found = len(found_options) == len(expected_options)
            self.log_result(test_id, test_name, 'Dropdown Options Verified', all_found,
                          f"Found {len(found_options)}/{len(expected_options)} options: {', '.join(found_options)}",
                          screenshot)

            return all_found

        except Exception as error:
            screenshot = self.take_screenshot(test_id, 'error')
            self.log_result(test_id, test_name, 'Test Execution', False,
                          f'Error: {str(error)}', screenshot)
            return False

    def test_case_4(self):
        """Test Case 4: Verify Create from Type Additional Options"""
        test_id = 'TC-MD-PS-001.4'
        test_name = 'Verify Additional Options on Create from Type'

        print(f"\n▶ Running {test_id}: {test_name}")

        try:
            # Step 1: Hover over Create from Type
            print("  Step 1: Hovering over Create from Type...")
            create_from_type = self.find_element_by_text('Create from Type')

            if not create_from_type:
                screenshot = self.take_screenshot(test_id, 'create-from-type-not-found')
                self.log_result(test_id, test_name, 'Create from Type Found', False,
                              'Could not locate Create from Type option', screenshot)
                return False

            self.hover_element(create_from_type)
            screenshot = self.take_screenshot(test_id, 'create-from-type-hovered')
            self.log_result(test_id, test_name, 'Create from Type Hovered', True,
                          'Hovered over Create from Type option', screenshot)

            # Step 2: Verify Purchased Part option
            print("  Step 2: Verifying Purchased Part option...")
            purchased_part = self.find_element_by_text('Purchased Part')
            screenshot = self.take_screenshot(test_id, 'submenu-options')

            if purchased_part:
                self.log_result(test_id, test_name, 'Purchased Part Option Found', True,
                              'Purchased Part option is visible', screenshot)
                return True
            else:
                self.log_result(test_id, test_name, 'Purchased Part Option Found', False,
                              'Purchased Part option not found', screenshot)
                return False

        except Exception as error:
            screenshot = self.take_screenshot(test_id, 'error')
            self.log_result(test_id, test_name, 'Test Execution', False,
                          f'Error: {str(error)}', screenshot)
            return False

    def test_case_5(self):
        """Test Case 5: Create Purchased Part and Verify Sub-tabs"""
        test_id = 'TC-MD-PS-001.5'
        test_name = 'Create Purchased Part and Verify Sub-tabs'

        print(f"\n▶ Running {test_id}: {test_name}")

        try:
            # Step 1: Click Purchased Part
            print("  Step 1: Clicking Purchased Part...")
            purchased_part = self.find_element_by_text('Purchased Part')

            if purchased_part:
                purchased_part.click()
                time.sleep(2)
                screenshot = self.take_screenshot(test_id, 'create-page-loaded')
                self.log_result(test_id, test_name, 'Create Page Loaded', True,
                              'Navigated to Create Purchased Part page', screenshot)
            else:
                screenshot = self.take_screenshot(test_id, 'purchased-part-not-found')
                self.log_result(test_id, test_name, 'Purchased Part Clicked', False,
                              'Could not locate or click Purchased Part', screenshot)
                return False

            # Step 2: Verify required sub-tabs
            print("  Step 2: Verifying sub-tabs...")
            required_subtabs = [
                'Key Info',
                'Product Management',
                'Planning & Estimating',
                'Cost & Prices',
                'Supplier Sources',
                'Comments & Files',
                'Texts'
            ]

            found_subtabs = self.find_elements_by_text(required_subtabs)
            screenshot = self.take_screenshot(test_id, 'subtabs-verification')

            all_found = len(found_subtabs) == len(required_subtabs)
            self.log_result(test_id, test_name, 'Sub-tabs Verified', all_found,
                          f"Found {len(found_subtabs)}/{len(required_subtabs)} sub-tabs: {', '.join(found_subtabs)}",
                          screenshot)

            return all_found

        except Exception as error:
            screenshot = self.take_screenshot(test_id, 'error')
            self.log_result(test_id, test_name, 'Test Execution', False,
                          f'Error: {str(error)}', screenshot)
            return False

    def run_all_tests(self):
        """Run all test cases"""
        print("=" * 50)
        print("Master Data - Products & Services Tests")
        print("=" * 50)
        print(f"Base URL: {BASE_URL}")
        print(f"Timestamp: {datetime.now().isoformat()}\n")

        try:
            self.init_driver()

            # Run test cases
            test1_result = self.test_case_1()
            test2_result = self.test_case_2()
            test3_result = self.test_case_3()
            test4_result = self.test_case_4()
            test5_result = self.test_case_5()

            # Generate report
            print("\n" + "=" * 50)
            print("Test Summary")
            print("=" * 50)
            print(f"Total Tests: {test_report['summary']['totalTests']}")
            print(f"Passed: {test_report['summary']['passed']}")
            print(f"Failed: {test_report['summary']['failed']}")
            if test_report['summary']['totalTests'] > 0:
                pass_rate = (test_report['summary']['passed'] / test_report['summary']['totalTests']) * 100
                print(f"Pass Rate: {pass_rate:.2f}%")

            # Save report
            with open(REPORT_FILE, 'w') as f:
                json.dump(test_report, f, indent=2)

            print(f"\nTest report saved to: {REPORT_FILE}")
            print(f"Screenshots saved to: {SCREENSHOT_DIR}")

        except Exception as error:
            print(f"Test execution failed: {error}")
        finally:
            if self.driver:
                self.driver.quit()


if __name__ == '__main__':
    test_automation = TestAutomation()
    test_automation.run_all_tests()
